### PR TITLE
Fix NullPointerException in ShieldExpansionEvents.getShieldValue

### LINKED ISSUE #47: Crash: Rendering screen
| Name                 | Version |
|-------------------|----------|
| Minecraft            | 1.20.1   |
| Minecraft Forge  | 47.3.12 |
| Shield Expansion | 1.1.7a   |

Crash Reports: 

[crash-2024-12-06_23.52.21-client.txt](https://github.com/user-attachments/files/18045204/crash-2024-12-06_23.52.21-client.txt)
[crash-2024-12-06_23.54.22-client.txt](https://github.com/user-attachments/files/18045201/crash-2024-12-06_23.54.22-client.txt)
[crash-2024-12-06_23.55.33-client.txt](https://github.com/user-attachments/files/18045199/crash-2024-12-06_23.55.33-client.txt)

For context, I'm playing on a server and it seems that I crash whenever entering a gui with a damaged shield on my hotbar - unsure whether the shield is broken somehow due to me taking it out of a Refined Storage system.

Offending code seems to be https://github.com/infernalstudios/Shield-Expansion/blob/029922574e272de62289198ff7752af4bf353adc/src/main/java/org/infernalstudios/shieldexp/events/ShieldExpansionEvents.java#L244-L247

So I'm assuming either `SHIELD_STATS.get(key)` or `SHIELD_STATS.get("shieldexp:default")` is returning null(?)

### PR DESCRIPTION
**Pull request by artman41**
Fixes #47

### CODE REVIEW

File: src/main/java/org/infernalstudios/shieldexp/events/ShieldExpansionEvents.java [hunk 1]
```diff
@@ -243,7 +245,15 @@ public boolean validateBlocking(Player player) {
     //reads a shield attribute from the given shield's stats map, or the default map if no map is found
     public static Double getShieldValue(Item item, String value) {
         String key = ForgeRegistries.ITEMS.getKey(item).toString();
-        return SHIELD_STATS.containsKey(key) ? SHIELD_STATS.get(key).get(value) : SHIELD_STATS.get("shieldexp:default").get(value);
+        Map<String, Double> statsMap = null;
+        if(SHIELD_STATS.containsKey(key))
+            statsMap = SHIELD_STATS.get(key);
+        if(statsMap == null)
+            SHIELD_STATS.get("shieldexp:default");
```
Comment (Jonathing) [inline src/main/java/org/infernalstudios/shieldexp/events/ShieldExpansionEvents.java]: You are not assigning this value to anything. What are you trying to do here?

**Review by Jonathing — CHANGES_REQUESTED** [review]

File: build.properties [hunk 2]
```diff
@@ -1,4 +1,4 @@
-VERSION=1.1.7a
+VERSION=1.1.7b
```
Comment (SwanX1) [inline build.properties]: Please do not change version numbers in pull requests.

File: src/main/java/org/infernalstudios/shieldexp/events/ShieldExpansionEvents.java (hunk 1 above)
Comment (artman41) [reply src/main/java/org/infernalstudios/shieldexp/events/ShieldExpansionEvents.java]: corrected the assignment 👍

File: build.properties (hunk 2 above)
Comment (artman41) [reply build.properties]: reverted the version 👍

**Comment by Jonathing** [comment]
The code looks fine now, but this doesn't solve the underlying issue that `SHIELD_STATS.get(key)` or `SHIELD_STATS.get("shieldexp:default")` could be null. It might be worth investigating into how shield stats are actually saved, or if there are any null values being saved into the map. Just because a map has a key doesn't mean that its value will not be null.

**Comment by artman41** [comment]
> The code looks fine now, but this doesn't solve the underlying issue that `SHIELD_STATS.get(key)` or `SHIELD_STATS.get("shieldexp:default")` could be null. It might be worth investigating into how shield stats are actually saved, or if there are any null values being saved into the map. Just because a map has a key doesn't mean that its value will not be null.

I'll drop some debug info into a local build and see if I can get an idea of which map is the issue

Any recommendations as to where to start looking for map population?

I realised earlier that another cause of the issue could be that I added the mod into an already-existing world where shields had already been crafted

**Comment by Jonathing** [comment]
I looked through it real quick and there is only one possibility that makes sense: the shields aren't properly synced to the client. This is going to be a bitch and a half to test, but it makes sense given how the synchronization packet is handled.

In the SyncShields packet, [SyncShields#handle()](https://github.com/infernalstudios/Shield-Expansion/blob/029922574e272de62289198ff7752af4bf353adc/src/main/java/org/infernalstudios/shieldexp/network/SyncShields.java#L41-L47) has a null-check for the player and does not synchronize shields if the player is null. While this makes sense, this packet is sent on `PlayerEvent.PlayerLoggedInEvent`, and it's entirely possible that the client hasn't finished setting up its own player.

The LocalPlayer is never actually accessed when writing to or clearing the local shield data, so you could probably remove the null-check without any side effects. If the packet is going through, then the data isn't corrupted. I do think the syncing of shields could be a lot cleaner, but it might address the issue of why the default shield data might be missing in the first place.

**Comment by Jonathing** [comment]
The alternative (less desirable) solution is to just hard-code some default shield stats for the client, but that wouldn't solve the potential syncing problem. After reading all usages of `SHIELD_STATS`, it isn't adding any null values to the map, so the null pointer is occurring at `SHIELD_STATS.get("shieldexp:default")`.

Try removing the player null-checks from both SyncShields and ClearShields and see if the problem persists after that. Because the solution this PR provides doesn't solve the bigger issue.

**Comment by SwanX1** [comment]
Superseded by #49

**Pull request closed without merging**